File: AEDs/Algoritmos de Ordenacao/C/BubbleBetter.c
#include "Geracao.h"
#include "Log.h"

void sort ( int* arranjo, int tamanho, Log* log )
{
    bool swapped = true;
    for( int i = 0; i < tamanho-1 && swapped; i++ )
    {
        swapped = false;
        for( int j = 0; j < tamanho-i-1; j++ )
        {
            incrementarComparacoes( log, 1 );
            if( arranjo[j] >arranjo[j+1] )
            {
                swap( &arranjo[j], &arranjo[j+1] );
                incrementarMovimentacoes( log, 3 );
                swapped = true;
            } // end if
        } // end for
    } // end for
} // end sort ( )

int main ( void )
{
    Geracao* arranjo = new_Geracao( 100 );
    Log*     log     = new_Log ( );

    printf( "\n//----- BUBBLE BETTER SORT -----//\n");

    preencher_aleatorio( arranjo, 200 );

    printf("\nARRANJO PRE ORDENACAO:\n");
    mostrar( arranjo );

    inicioTempo( log );
    sort( arranjo->arranjo, arranjo->tamanho, log );
    fimTempo( log );

    printf("\nARRANJO POS ORDENACAO:\n");
    mostrar( arranjo );

    registrar ( log );

    delete_Log ( log );
    delete_Geracao( arranjo );
} // end main ( )

File: AEDs/Algoritmos de Ordenacao/Java/BubbleBetter.java
public class BubbleBetter extends Geracao
{
    public BubbleBetter ( )
    {
        super ( );
    }

    public BubbleBetter ( int tamanho )
    {
        super( tamanho );
    }

    /**
     *  <p>Bubble Sort é um algoritmo de ordenação simples, que percorre repetidamente
     *  a lista, compara elementos adjacentes e os troca se estiverem na ordem
     *  errada.</p>
     *  
     *  <p>Complexidade: O(n^2) no pior caso, O(n) no melhor caso. </p>
     *  <p>Número de comparações: O(n^2).                          </p>
     *  <p>Número de movimentações: O(n^2).                        </p>
     */
    @Override
    public void sort ( Log log )
    {
        boolean swapped = true;
        for( int i = 0; i < tamanho-1 && swapped; i++ )
        {
            swapped = false;
            for( int j = 0; j < tamanho-i-1; j++ )
            {
                log.incrementarComparacoes();
                if( arranjo[j] > arranjo[j+1] )
                {
                    swap( j, j+1 );
                    log.incrementarMovimentacoes(3);
                    swapped = true;
                } // end if
            } // end for
        } // end for
    } // end sort ( )

    public static void main( String[] args ) 
    {
        BubbleBetter arranjo = new BubbleBetter( 100 );
        Log log = new Log( );

        System.out.println( "\n//----- BUBBLE BETTER SORT -----//");

        arranjo.preencher_aleatorio();

        System.out.println("\nARRANJO PRE ORDENACAO:");
        arranjo.mostrar( );

        log.inicioTempo();
        arranjo.sort( log );
        log.fimTempo();

        System.out.println("\nARRANJO POS ORDENACAO:");
        arranjo.mostrar( );

        log.mostrar();
    } // end main ( )
} // end class

